fdbserver/RestoreMaster.actor.cpp

```diff
@@ -675,6 +712,78 @@ ACTOR static Future<Version> collectBackupFiles(Reference<IBackupContainer> bc,
 	return request.targetVersion;
 }
 
+// By the first and last block of *file to get (beginKey, endKey);
+// set (beginKey, endKey) and file->version to pRangeVersions
+ACTOR static Future<Void> insertRangeVersion(KeyRangeMap<Version>* pRangeVersions, RestoreFileFR* file,
+                                             Reference<IBackupContainer> bc) {
+	TraceEvent("FastRestoreMasterDecodeRangeVersion").detail("File", file->toString());
+	state Reference<IAsyncFile> inFile = wait(bc->readFile(file->fileName));
+	state bool beginKeySet = false;
+	state Key beginKey;
+	state Key endKey;
+	state int64_t j = 0;
+	for (; j < file->fileSize; j += file->blockSize) {
+		int64_t len = std::min<int64_t>(file->blockSize, file->fileSize - j);
+		Standalone<VectorRef<KeyValueRef>> blockData = wait(parallelFileRestore::decodeRangeFileBlock(inFile, j, len));
+		if (!beginKeySet) {
+			beginKey = blockData.front().key;
+		}
+		endKey = blockData.back().key;
+	}
+
+	// First and last key are the range for this file: endKey is exclusive
+	KeyRange fileRange = KeyRangeRef(beginKey.contents(), endKey.contents());
+	TraceEvent("FastRestoreMasterInsertRangeVersion")
+	    .detail("DecodedRangeFile", file->fileName)
+	    .detail("KeyRange", fileRange)
+	    .detail("Version", file->version);
+	// Update version for pRangeVersions's ranges in fileRange
+	auto ranges = pRangeVersions->modify(fileRange);
+	for (auto r = ranges.begin(); r != ranges.end(); ++r) {
+		r->value() = std::max(r->value(), file->version);
+	}
+
+	// Dump the new key ranges
+	ranges = pRangeVersions->ranges();
+	int i = 0;
+	for (auto r = ranges.begin(); r != ranges.end(); ++r) {
+		TraceEvent(SevDebug, "RangeVersionsAfterUpdate")
+		    .detail("File", file->toString())
+		    .detail("FileRange", fileRange.toString())
+		    .detail("FileVersion", file->version)
+		    .detail("RangeIndex", i++)
+		    .detail("RangeBegin", r->begin())
+		    .detail("RangeEnd", r->end())
+		    .detail("RangeVersion", r->value());
+	}
+
+	return Void();
+}
+
+// Build the version skyline of snapshot ranges by parsing range files;
+// Expensive and slow operation that should not run in real prod.
+ACTOR static Future<Void> buildRangeVersions(KeyRangeMap<Version>* pRangeVersions,
+                                             std::vector<RestoreFileFR>* pRangeFiles, Key url) {
+	if (!g_network->isSimulated()) {
+		TraceEvent(SevError, "ExpensiveBuildRangeVersions")
+		    .detail("Reason", "Parsing all range files is slow and memory intensive");
+		return Void();
+	}
+	Reference<IBackupContainer> bc = IBackupContainer::openContainer(url.toString());
+
+	// Key ranges not in range files are empty;
+	// Assign highest version to avoid applying any mutation in these ranges
+	state int fileIndex = 0;
+	state std::vector<Future<Void>> fInsertRangeVersions;
```

Comment:
Nit. Both can be "non-state" variables and `fileIndex` can be inlined in the for loop.